src/happy.rs

```diff
@@ -0,0 +1,84 @@
+//! aaa
+use std::io;
+use std::iter::FusedIterator;
+use std::net::TcpStream;
+use std::net::ToSocketAddrs;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::mpsc::channel;
+use std::sync::Arc;
+use std::thread;
+use std::time::{Duration, Instant};
+
+fn intertwine<T, A, B>(mut ita: A, mut itb: B) -> Vec<T>
+where
+    A: FusedIterator<Item = T>,
+    B: FusedIterator<Item = T>,
+{
+    let mut res = vec![];
+
+    loop {
+        match (ita.next(), itb.next()) {
+            (Some(a), Some(b)) => {
+                res.push(a);
+                res.push(b);
+            }
+            (Some(a), None) => {
+                res.push(a);
+            }
+            (None, Some(b)) => {
+                res.push(b);
+            }
+            (None, None) => break,
+        }
+    }
+
+    res
+}
+
+/// aaa
+pub fn connect<A>(addrs: A) -> io::Result<TcpStream>
+where
+    A: ToSocketAddrs,
+{
+    let addrs: Vec<_> = addrs.to_socket_addrs()?.collect();
+    let ipv4 = addrs.iter().filter(|a| a.is_ipv4()).cloned();
+    let ipv6 = addrs.iter().filter(|a| a.is_ipv6()).cloned();
+    let order = intertwine(ipv4, ipv6);
+
+    let done = Arc::new(AtomicBool::new(false));
```

Comment:
I wonder if the additional synchronization via `done` and the intermediate thread to spawn the connection threads is really helpful. Couldn't we remove this and replace the `thread::sleep` with a `recv_timeout` call on `rx`?

Reply:
Yes that's a good idea. It would avoid spawning a thread for the loop as well. I've also been looking at threadpool libraries, I'm not sure if the performance cost of spawning threads like this is big enough to warrant using a thread pool library.

Reply:
Personally, I would not go for a thread pool here to keep things simple. It might make sense to limit the number of threads spawned though, e.g. try to connect to at most `N` addresses in parallel by effectively throttling the loop.

Reply:
I did a few experiments, nothing very thorough but I usually don't see a ton of addresses attached to a domain, so I think that for now we can leave it like this and limit the number of threads if problems occur.

Reply:
Agreed.